### ISSUE TITLE
[Request] Name generated header after module name if provided

### ISSUE THREAD

**Issue by thii** [bazelbuild/rules_swift#346]
Currently the generated header of a `swift_library` is named after the target name. I'd like to request to change it to be named after the `module_name` attribute if provided, and `name` otherwise.

Our use case:
We build a mixed Obj-C and Swift static framework by building a `swift_library` target with the `-import-underlying-module` flag, and use an `objc_library` target as the wrapper target, with both target share the same module name. Since a target name has to be unique within a package, we name the `swift_library` target after the module name (for instance `MyLibrary`) in order to get the generated header named `MyLibrary-Swift.h` so that we can import it from the Obj-C part using `#import <MyLibrary/MyLibrary-Swift.h>`. Then for the wrapper target, we're having to name it something else (`MyLibrary_umbrella`) in order to make it work.

The build command for the wrapper target would be:
```
bazel build //path/to/MyLibrary:MyLibrary_umbrella
```

and the `deps` declaration for other targets that depend on this would be:

```
deps = [
    "//path/to/MyLibrary:MyLibrary_umbrella"
]
```

With this change, the above build command and `deps` declaration would be:

```
bazel build //path/to/MyLibrary
```

```
deps = [
    "//path/to/MyLibrary"
]
```

**Comment by allevato** [comment]
This would be a breaking change, so it's not something we could do lightly—at a minimum, it could require some amount of migration for our internal users if they have any libraries with custom module names used from Objective-C (though we try to discourage using the `module_name` attribute entirely).

Doing this would also make the name of the generated header less predictable to the end user. Today, if you have a target `//foo/bar:baz`, then you know your Objective-C code can `#import "foo/bar/baz-Swift.h"`. Making that conditional on the module name makes that behavior harder to explain and document.

Ultimately I think the issue here isn't one that should be solved by renaming the generated header, but by having first-class mixed Swift/Objective-C support in the same module so that these kinds of hacks aren't necessary.

**Comment by thii** [comment]
Thanks for the detailed explanation. That totally makes sense to me. Is first-class mixed Swift/Objective-C support in the same module in the roadmap 🙂?

For the time being, would it be possible to loose the spec a bit by allowing a custom name for the generated header via a `--features` flag?

**Comment by mpalczew** [comment]
or how about just adding a new field to override the header name.

**Comment by segiddins** [comment]
Being able to customize the header name (inline with what Xcode allows with `SWIFT_OBJC_INTERFACE_HEADER_NAME`) would make migrating a codebase to support building on Bazel while still building in Xcode significantly easier, otherwise we need to use `#ifdef`s or `#if __can_include()`s

**Comment by allevato** [comment]
Of the options presented here, having an attribute like `generated_header_name` is probably the least objectionable. We can document what the default is if it's not specified and there's no mysterious behavior—users don't have to guess what the name will be based on other feature flags (which don't even have to be specified locally on the build target; they could be inherited from the package or the command line).

**Comment by segiddins** [comment]
Would a PR adding `generated_header_name` be a good way forward?

**Comment by allevato** [comment]
It would be. (And you have!)

**Comment by thii** [comment]
Resolved by #385.
